alf/algorithms/ppg/disjoint_policy_value_network.py

```diff
@@ -0,0 +1,250 @@
+            continuous_projection_net_ctor=continuous_projection_net_ctor,
+            input_size=self._actor_encoder.output_spec.shape[0],
+            action_spec=action_spec)
+        # In the case when the projection net is multi-fold (because actor
+        # spec is a nest), force picking up the parameters inside those
+        # project networks into this DisjointPolicyValueNetwork instance.
+        self._projection_net_module_list = torch.nn.ModuleList(
+            alf.nest.flatten(self._policy_head))
+
+        # +------------------------------------------+
+        # | Step 3: Value head of the aux branch     |
+        # +------------------------------------------+
+
+        # Note that the aux branch value head belongs to the policy component.
+
+        # Like the value head Aux head is outputing value estimation
+        self._aux_head = alf.nn.Sequential(
+            alf.layers.FC(
+                input_size=self._actor_encoder.output_spec.shape[0],
+                output_size=1), alf.layers.Reshape(shape=()))
+
+        # +------------------------------------------+
+        # | Step 4: Assemble network + value head    |
+        # +------------------------------------------+
+
+        if is_sharing_encoder:
+            # Use the same encoder, but the encoder is DETACHED.
+            self._value_head = alf.nn.Sequential(
+                alf.layers.Detach(),
+                alf.layers.FC(
+                    input_size=self._actor_encoder.output_spec.shape[0],
+                    output_size=1),
+                alf.layers.Reshape(shape=()),
+                input_tensor_spec=self._actor_encoder.output_spec)
+
+            all_branches = ((self._policy_head, self._aux_head),
+                            self._value_head)
+            self._composition = alf.nn.Sequential(
+                self._actor_encoder,
+                alf.nn.Branch(all_branches, name='AllBranches'))
+        else:
+            policy_component = alf.nn.Sequential(
+                self._actor_encoder,
+                alf.nn.Branch((self._policy_head, self._aux_head),
+                              name='PolicyComponent'))
+
+            # When not sharing encoder, create a separate encoder for the value
+            # component.
+            self._value_encoder = encoding_network_ctor(
+                input_tensor_spec=observation_spec,
+                kernel_initializer=kernel_initializer)
+
+            self._value_head = alf.nn.Sequential(
+                alf.layers.FC(
+                    input_size=self._actor_encoder.output_spec.shape[0],
+                    output_size=1), alf.layers.Reshape(shape=()))
+
+            value_component = alf.nn.Sequential(self._value_encoder,
+                                                self._value_head)
+
+            self._composition = alf.nn.Branch(
+                (policy_component, value_component), name='AllBranches')
```

Comment:
It's possible to make the network return the right structure as the following. And we don't even need to implement forward. Simply return the constructed network is enough. And network state will correctly handled by the container, which means that RNN encoder is automatically supported.
```python
    self._composition = alf.nn.Sequential(
        alf.nn.Branch(
                (policy_component, value_component), name='AllBranches'),
        lambda heads: (heads[0][0], heads[1], heads[0][1]))
```

Reply:
Do you think a single giant statement like the following is easier to understand?
```python
    composition = alf.nn.Sequential(
        alf.nn.Branch(
            alf.nn.Sequential(
                self._actor_encoder,
                alf.nn.Branch(self._policy_head, self._aux_head)),
            alf.nn.Sequential(
                self._value_encoder,
                alf.layers.FC(input_size=encoder_out_size, output_size=1),
                alf.layers.Reshape(shape=()))),
        lambda heads: (heads[0][0], heads[1], heads[0][1]))
```